File: tools/pageless-pdf/vector-compare.ts
#!/usr/bin/env node
// SPDX-License-Identifier: AGPL-3.0-or-later
// Exact placement: the strip's glyphs and rules against the viewer's.
//
//     node tools/pageless-pdf/vector-compare.ts <build dir | pageless.pdf> <url> [--out DIR] [--margin PT]
//                                               [--window PT] [--line-tol PT] [--wait-log TEXT]
//
// No pixels. MuPDF (in WebAssembly) reads every glyph LuaTeX wrote into the
// strip with its origin, and dom-dump.ts every glyph the viewer put on the page
// (tspan x/y) in the same frame: pt from the column's margin edge, pt from the
// top. Each viewer glyph is matched to the nearest strip glyph (within
// --window; no character matching: a CM glyph's text differs between the two),
// and the residuals say where the viewer's geometry departs from TeX's, to a
// hundredth of a point:
//
//  - vertical: after aligning the two at the top, the browser's height as a
//    function of position (a drift: its stacking loses or gains height);
//  - horizontal: per text line, the offset at its start (protrusion,
//    indentation) and the slope along it (a glue or expansion ratio that
//    differs from TeX's);
//  - rules: the viewer's rects against the strip's rules, corner by corner, at
//    any angle.
//
// Writes vector.json (every matched glyph's residual, the per-line table, the
// rule table) and prints the lines that are off. Run it before compare.ts:
// what it reports is geometry, not rasterisation.
import { parseArgs } from 'node:util';
import { existsSync, mkdirSync, readFileSync, statSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import * as mupdf from 'mupdf';
import { dumpDom, type Dom } from './dom-dump.ts';

const PT_PER_BP = 72.27 / 72;
type Pt = [number, number];

/** The strip's glyph origins (pt, y down; a ligature's text-only characters
 *  left out) and its rules as four corners each. */
export function traceStrip(pdf: string): { glyphs: Pt[]; rules: Pt[][] } {
  const page = mupdf.Document.openDocument(readFileSync(pdf), 'application/pdf').loadPage(0);
  const glyphs: Pt[] = [], rules: Pt[][] = [];
  const mul = (m: number[], x: number, y: number): Pt => [(m[0] * x + m[2] * y + m[4]) * PT_PER_BP, (m[1] * x + m[3] * y + m[5]) * PT_PER_BP];
  const points = (path: mupdf.Path) => {
    const pts: [number, number][] = [];
    path.walk({ moveTo: (x, y) => pts.push([x, y]), lineTo: (x, y) => pts.push([x, y]) });
    return pts;
  };
  const device = new mupdf.Device({
    fillText(text: mupdf.Text, ctm: mupdf.Matrix) {
      text.walk({ showGlyph(_font: mupdf.Font, trm: mupdf.Matrix, gid: number) { if (gid >= 0) glyphs.push(mul(ctm as number[], trm[4], trm[5])); } });
    },
    // LuaTeX strokes a rule along its centre line, lw thick (butt caps; a
    // square cap reaches lw/2 further), horizontal or vertical as the rule is
    // wide or tall, and under \rotatebox through a rotating transform; a filled
    // four-cornered path is taken as it stands.
    strokePath(path: mupdf.Path, stroke: mupdf.StrokeState, ctm: mupdf.Matrix) {
      const pts = points(path);
      if (pts.length !== 2) return;
      let [[x0, y0], [x1, y1]] = pts;
      const len = Math.hypot(x1 - x0, y1 - y0);
      if (len === 0) return;
      const ux = (x1 - x0) / len, uy = (y1 - y0) / len, hw = stroke.getLineWidth() / 2;
      const cap = stroke.getLineCap() !== 0 ? hw : 0;
      x0 -= ux * cap; y0 -= uy * cap; x1 += ux * cap; y1 += uy * cap;
      const nx = -uy * hw, ny = ux * hw;
      rules.push([[x0 + nx, y0 + ny], [x1 + nx, y1 + ny], [x1 - nx, y1 - ny], [x0 - nx, y0 - ny]].map(([x, y]) => mul(ctm as number[], x, y)));
    },
    fillPath(path: mupdf.Path, _evenOdd: boolean, ctm: mupdf.Matrix) {
      let pts = points(path);
      if (pts.length === 5 && pts[0][0] === pts[4][0] && pts[0][1] === pts[4][1]) pts = pts.slice(0, 4);
      if (pts.length === 4) rules.push(pts.map(([x, y]) => mul(ctm as number[], x, y)));
    },
  });
  page.run(device, mupdf.Matrix.identity);
  device.close();
  return { glyphs, rules };
}

// ── numpy's few operations ──────────────────────────────────────────────────
const mean = (a: number[]) => a.reduce((s, x) => s + x, 0) / a.length;
const std = (a: number[]) => { const m = mean(a); return Math.sqrt(mean(a.map(x => (x - m) ** 2))); };
const median = (a: number[]) => { const s = [...a].sort((x, y) => x - y), n = s.length; return n % 2 ? s[(n - 1) / 2] : (s[n / 2 - 1] + s[n / 2]) / 2; };
/** first index i with s[i] >= v (numpy.searchsorted, side='left') */
const searchsorted = (s: number[], v: number) => { let lo = 0, hi = s.length; while (lo < hi) { const m = (lo + hi) >> 1; if (s[m] < v) lo = m + 1; else hi = m; } return lo; };
/** the slope of the least-squares line (numpy.polyfit(x, y, 1)[0]) */
const slopeOf = (x: number[], y: number[]) => { const mx = mean(x), my = mean(y); let n = 0, d = 0; x.forEach((xi, i) => { n += (xi - mx) * (y[i] - my); d += (xi - mx) ** 2; }); return n / d; };
const round = (x: number, k: number) => Math.round(x * 10 ** k) / 10 ** k;

export interface CompareOptions { out?: string; margin?: number; window?: number; lineTol?: number; waitLog?: string; log?: (s: string) => void }

export async function vectorCompare(build: string, url: string, o: CompareOptions = {}): Promise<Record<string, unknown>> {
  const log = o.log ?? (s => console.log(s));
  const win = o.window ?? 0.7, lineTol = o.lineTol ?? 0.1;
  const pdf = existsSync(build) && statSync(build).isDirectory() ? join(build, 'pageless.pdf') : build;
  const meta = JSON.parse(readFileSync(join(dirname(pdf), 'pageless.json'), 'utf8'));
  const sp = meta.sp_per_pt, hsize = meta.hsize / sp, stripMargin = meta.margin / sp;
  const margin = o.margin ?? stripMargin;
  const out = o.out ?? join(dirname(pdf), 'vector');
  mkdirSync(out, { recursive: true });

  const { glyphs: P, rules } = traceStrip(pdf);
  const dom: Dom = await dumpDom(url, { hsize, margin, waitLog: o.waitLog });
  writeFileSync(join(out, 'dom.json'), JSON.stringify(dom));
  const G = dom.glyphs;
  log(`strip: ${P.length} glyphs, ${rules.length} rules; viewer: ${G.length} glyphs, ${dom.rects.length} rects`);

  // the viewer's frame starts `margin` left of the column; the strip's `stripMargin`
  const shift = stripMargin - margin;
  const D = G.map(g => [g.x + shift, g.y] as Pt);
  const order = P.map((_, i) => i).sort((a, b) => P[a][1] - P[b][1]);
  const Py = order.map(i => P[i][1]);
  const matched: [number, number, number, number][] = [];            // (dom index, strip index, dx, dy), strip − viewer
  G.forEach((g, j) => {
    if (!g.text.trim()) return;
    const ty = D[j][1];
    const lo = searchsorted(Py, ty - win), hi = searchsorted(Py, ty + win);
    let best = -1, bestD = Infinity;
    for (let c = lo; c < hi; c++) {
      const i = order[c], dx = Math.abs(P[i][0] - D[j][0]), dy = Math.abs(P[i][1] - D[j][1]);
      if (dx < win && dx + dy < bestD) { bestD = dx + dy; best = i; }
    }
    if (best >= 0) matched.push([j, best, P[best][0] - D[j][0], P[best][1] - D[j][1]]);
  });
  const nInk = G.filter(g => g.text.trim()).length;
  log(`matched ${matched.length} of ${nInk} viewer glyphs within ${win} pt`);
  if (!matched.length) throw new Error('nothing matched: are the two the same document at the same width?');
  const dys = matched.map(m => m[3]), ys = matched.map(m => G[m[0]].y);

  // vertical: the drift, as the median dy per 250 pt of height
  const drift: (number | null)[] = [];
  for (let e0 = 0; e0 < Math.max(...ys) + 250 - 250; e0 += 250) {
    const s = dys.filter((_, k) => ys[k] >= e0 && ys[k] < e0 + 250);
    drift.push(s.length > 10 ? round(median(s), 3) : null);
  }
  const shown = drift.filter(d => d !== null);
  log(`vertical (strip − viewer, pt): mean ${mean(dys).toFixed(3)}, sd ${std(dys).toFixed(3)}, largest ${Math.max(...dys.map(Math.abs)).toFixed(3)}; ` +
      `drift per 250 pt: ${JSON.stringify(shown.slice(0, 8))} … ${JSON.stringify(shown.slice(-3))}`);

  // horizontal, per line (glyphs sharing a baseline): offset at the start, slope along it
  const lines = new Map<number, [number, number, string][]>();
  for (const [j, , dx] of matched) {
    const y = round(G[j].y, 2);
    if (!lines.has(y)) lines.set(y, []);
    lines.get(y)!.push([D[j][0], dx, G[j].text]);
  }
  const table = [...lines].sort((a, b) => a[0] - b[0]).map(([y, pts]) => {
    pts.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    const x = pts.map(p => p[0]), dx = pts.map(p => p[1]);
    const slope = pts.length >= 6 && Math.max(...x) - Math.min(...x) > 50 ? slopeOf(x, dx) * 100 : 0;
    return { y, glyphs: pts.length, start_dx: round(dx[0], 3), end_dx: round(dx[dx.length - 1], 3),
             max_abs_dx: round(Math.max(...dx.map(Math.abs)), 3), slope_pt_per_100pt: round(slope, 3), text: pts.map(p => p[2]).join('').slice(0, 60) };
  });
  const off = table.filter(t => t.max_abs_dx > lineTol);
  log(`horizontal: ${table.length} lines, ${off.length} with a glyph off by more than ${lineTol} pt`);
  for (const t of [...off].sort((a, b) => b.max_abs_dx - a.max_abs_dx).slice(0, 25))
    log(`  y ${t.y.toFixed(2).padStart(9)}  start ${t.start_dx.toFixed(2)}  end ${t.end_dx.toFixed(2)}  slope ${t.slope_pt_per_100pt.toFixed(3)}/100pt  ${JSON.stringify(t.text)}`);

  // rules: each viewer rect against the strip rule whose corners are nearest.
  // `off` is the farthest any corner of either is from the other's nearest
  // corner (pt): position, size and angle in one number; dx/dy the move
  // between their centres (strip − viewer).
  const cornerDistance = (A: Pt[], B: Pt[]) => {
    const near = (p: Pt, Q: Pt[]) => Math.min(...Q.map(q => Math.hypot(p[0] - q[0], p[1] - q[1])));
    return Math.max(Math.max(...A.map(p => near(p, B))), Math.max(...B.map(q => near(q, A))));
  };
  const centre = (Q: Pt[]): Pt => [mean(Q.map(q => q[0])), mean(Q.map(q => q[1]))];
  const centres = rules.map(centre);
  const ruleRows: Record<string, unknown>[] = [], drawn = new Set<number>();
  for (const r of dom.rects) {
    if (r.w <= 0 || r.h <= 0) continue;
    const V = r.pts.map(([x, y]) => [x + shift, y] as Pt);
    const c = centre(V);
    const ptp = Math.max(Math.max(...V.map(v => v[0])) - Math.min(...V.map(v => v[0])), Math.max(...V.map(v => v[1])) - Math.min(...V.map(v => v[1])));
    let bestD = Infinity, k = -1;
    for (let i = 0; i < centres.length; i++) {
      if (Math.max(Math.abs(centres[i][0] - c[0]), Math.abs(centres[i][1] - c[1])) >= 2 + ptp / 2) continue;
      const d = cornerDistance(V, rules[i]);
      if (d < bestD) { bestD = d; k = i; }
    }
    if (k >= 0 && bestD < 2) {
      const d = bestD;
      drawn.add(k);
      ruleRows.push({ y: round(r.y, 2), x: round(r.x + shift, 2), off: round(d, 3), dx: round(centres[k][0] - c[0], 3), dy: round(centres[k][1] - c[1], 3) });
    } else ruleRows.push({ y: round(r.y, 2), unmatched: true, w: round(r.w, 2), h: round(r.h, 2) });
  }
  // and the rules TeX drew that the browser did not
  const missing = rules.map((q, k) => [q, k] as const).filter(([, k]) => !drawn.has(k))
    .map(([q]) => ({ y: round(Math.min(...q.map(p => p[1])), 2), x: round(Math.min(...q.map(p => p[0])), 2) }));
  const rm = ruleRows.filter(r => 'dx' in r) as { dx: number; dy: number; off: number; y: number }[];
  if (rm.length) {
    log(`rules: ${rm.length} matched of ${ruleRows.length} drawn, ${missing.length} of TeX's not drawn; strip − viewer x sd ${std(rm.map(r => r.dx)).toFixed(3)}, ` +
        `y sd ${std(rm.map(r => r.dy)).toFixed(3)}, largest corner distance ${Math.max(...rm.map(r => r.off)).toFixed(3)} pt`);
    for (const r of rm.filter(r => r.off > lineTol).sort((a, b) => b.off - a.off).slice(0, 10))
      log(`  rule at y ${r.y.toFixed(2).padStart(9)}: corners ${r.off.toFixed(2)} pt off, centre dx ${r.dx.toFixed(2)} dy ${r.dy.toFixed(2)} pt`);
  }
  const result = {
    hsize_pt: hsize, margin_pt: margin, window_pt: win,
    glyphs: { strip: P.length, viewer: nInk, matched: matched.length },
    vertical: { mean: round(mean(dys), 4), sd: round(std(dys), 4), max_abs: round(Math.max(...dys.map(Math.abs)), 4), drift_per_250pt: drift },
    lines: table, lines_off: off, rules: ruleRows, rules_missing: missing,
    matched: matched.map(([j, , dx, dy]) => ({ y: round(G[j].y, 2), x: round(D[j][0], 2), text: G[j].text, font: G[j].font, dx: round(dx, 3), dy: round(dy, 3) })),
  };
  writeFileSync(join(out, 'vector.json'), JSON.stringify(result, null, 1));
  log(`→ ${join(out, 'vector.json')}`);
  return result;
}

if (import.meta.main) {
  const { values: v, positionals } = parseArgs({
    allowPositionals: true,
    options: { out: { type: 'string' }, margin: { type: 'string' }, window: { type: 'string', default: '0.7' },
               'line-tol': { type: 'string', default: '0.1' }, 'wait-log': { type: 'string' } },
  });
  if (positionals.length !== 2) { console.error('usage: vector-compare.ts <build dir | pageless.pdf> <url> [--out DIR] [--margin PT] [--window PT] [--line-tol PT] [--wait-log TEXT]'); process.exit(2); }
  try {
    await vectorCompare(positionals[0], positionals[1], { out: v.out, margin: v.margin === undefined ? undefined : Number(v.margin),
      window: Number(v.window), lineTol: Number(v['line-tol']), waitLog: v['wait-log'] });
  } catch (e) { console.error(`error: ${(e as Error).message}`); process.exit(1); }
}

File: tools/pageless-pdf/vector_compare.py
#!/usr/bin/env python3
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Exact placement: the strip's glyphs and rules against the viewer's.

    vector_compare.py <build dir | pageless.pdf> <url> [--out DIR] [--margin PT]

No pixels. MuPDF's trace (`mutool trace`) gives every glyph LuaTeX wrote
into the strip with its origin, and dom_dump.js gives every glyph the
viewer put on the page (tspan x/y) in the same frame: pt from the column's
margin edge, pt from the top. Each viewer glyph is matched to the nearest
strip glyph (within --window, no character matching: a CM glyph's unicode
differs between the two), and the residuals say where the viewer's
geometry departs from TeX's, to a hundredth of a point:

  - vertical: after the two blocks are aligned at the top, the browser's
    height as a function of position (a drift means the browser's stacking
    loses or gains height);
  - horizontal: per text line, the offset at the line's start (protrusion
    or indentation differences) and the slope along the line (a glue or
    expansion ratio that differs from TeX's);
  - rules: the viewer's rects against the strip's rules, corner by corner,
    at any angle.

Writes vector.json (every matched glyph's residual, the per-line table,
the rule table) and prints the lines that are off. This is the check to
run before compare.py: whatever it reports is geometry, not rasterisation.
"""
from __future__ import annotations
import argparse
import collections
import json
import os
import re
import subprocess
import sys
from pathlib import Path

import numpy as np

HERE = Path(__file__).resolve().parent
PT_PER_BP = 72.27 / 72


def trace_strip(pdf: Path, out: Path) -> tuple[list, list]:
    """Glyphs [(x, y, unicode, font, size)] and rules [4×2 array of corners]
    of the strip's single page, in pt, y from the top."""
    txt = subprocess.run(['mutool', 'trace', str(pdf), '1'], capture_output=True, text=True, check=True).stdout
    out.write_text(txt)
    glyphs, rules = [], []
    for m in re.finditer(r'<fill_text[^>]*transform="([^"]+)">(.*?)</fill_text>', txt, re.S):
        a, b, c, d, e, f = [float(v) for v in m.group(1).split()]
        for sp in re.finditer(r'<span font="([^"]+)"[^>]*trm="([^"]+)">(.*?)</span>', m.group(2), re.S):
            font = sp.group(1).split('+')[-1]
            trm = [float(v) for v in sp.group(2).split()]
            for g in re.finditer(r'<g unicode="([^"]*)" glyph="[^"]*" x="([^"]+)" y="([^"]+)"', sp.group(3)):
                x, y = float(g.group(2)), float(g.group(3))
                glyphs.append(((a * x + c * y + e) * PT_PER_BP, (b * x + d * y + f) * PT_PER_BP,
                               g.group(1), font, trm[3] * PT_PER_BP, trm[0] / trm[3] if trm[3] else 1.0))
    # Rules: the four corners of what each path paints, at any angle. LuaTeX
    # strokes a rule along its centre line, lw thick (butt caps; a square
    # cap reaches lw/2 further), horizontal or vertical as the rule is wide
    # or tall, and under \rotatebox through a rotating transform; a filled
    # four-cornered path is taken as it stands.
    for m in re.finditer(r'<(stroke|fill)_path ([^>]*)>(.*?)</(?:stroke|fill)_path>', txt, re.S):
        attrs = dict(re.findall(r'(\w+)="([^"]*)"', m.group(2)))
        a, b, c, d, e, f = [float(v) for v in attrs['transform'].split()]
        pts = [(float(p.group(1)), float(p.group(2)))
               for p in re.finditer(r'<(?:moveto|lineto) x="([^"]+)" y="([^"]+)"', m.group(3))]
        if m.group(1) == 'stroke':
            if len(pts) != 2:
                continue
            (x0, y0), (x1, y1) = pts
            length = np.hypot(x1 - x0, y1 - y0)
            if length == 0:
                continue
            ux, uy = (x1 - x0) / length, (y1 - y0) / length
            hw = float(attrs['linewidth']) / 2
            cap = hw if attrs.get('linecap', '0').split(',')[0] != '0' else 0.0
            x0, y0, x1, y1 = x0 - ux * cap, y0 - uy * cap, x1 + ux * cap, y1 + uy * cap
            nx, ny = -uy * hw, ux * hw
            pts = [(x0 + nx, y0 + ny), (x1 + nx, y1 + ny), (x1 - nx, y1 - ny), (x0 - nx, y0 - ny)]
        else:
            if len(pts) == 5 and pts[0] == pts[4]:
                pts = pts[:4]
            if len(pts) != 4:
                continue
        rules.append(np.array([((a * x + c * y + e) * PT_PER_BP, (b * x + d * y + f) * PT_PER_BP) for x, y in pts]))
    return glyphs, rules


def dump_dom(url: str, out: Path, hsize: float, margin: float, node: str, wait_log: str | None = None) -> dict:
    cmd = [node, str(HERE / 'dom_dump.js'), url, str(out), '--hsize', repr(hsize), '--margin', repr(margin)]
    if wait_log:
        cmd += ['--wait-log', wait_log]
    r = subprocess.run(cmd, capture_output=True, text=True)
    sys.stderr.write(r.stderr)
    if r.returncode != 0:
        sys.exit(f'dom_dump.js failed ({r.returncode}): {r.stdout}')
    return json.loads(out.read_text())


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument('build', type=Path)
    ap.add_argument('url')
    ap.add_argument('--out', type=Path, default=None)
    ap.add_argument('--margin', type=float, default=None, help='white either side of the column, pt (default: the strip\'s)')
    ap.add_argument('--window', type=float, default=0.7, help='largest distance (pt) between a viewer glyph and the strip glyph it is taken to be')
    ap.add_argument('--line-tol', type=float, default=0.1, help='a line is reported when a glyph on it is off by more than this (pt)')
    ap.add_argument('--node', default='node')
    ap.add_argument('--wait-log', default=None, help='before dumping, wait for a console line containing this text')
    args = ap.parse_args()

    pdf = args.build / 'pageless.pdf' if args.build.is_dir() else args.build
    meta = json.loads(pdf.with_name('pageless.json').read_text())
    sp = meta['sp_per_pt']
    hsize, strip_margin = meta['hsize'] / sp, meta['margin'] / sp
    margin = strip_margin if args.margin is None else args.margin
    out = args.out or pdf.parent / 'vector'
    out.mkdir(parents=True, exist_ok=True)

    P, rules = trace_strip(pdf, out / 'trace.txt')
    dom = dump_dom(args.url, out / 'dom.json', hsize, margin, args.node, args.wait_log)
    G = dom['glyphs']
    print(f'strip: {len(P)} glyphs, {len(rules)} rules; viewer: {len(G)} glyphs, {len(dom["rects"])} rects')

    # the viewer's frame starts `margin` left of the column; the strip's `strip_margin`
    shift = strip_margin - margin
    Pxy = np.array([(g[0], g[1]) for g in P]) if P else np.zeros((0, 2))
    D = np.array([(g['x'] + shift, g['y']) for g in G]) if G else np.zeros((0, 2))
    order = np.argsort(Pxy[:, 1]); Py = Pxy[order, 1]
    matched = []                       # (dom index, strip index, dx, dy) with dx/dy = strip − viewer
    for j, g in enumerate(G):
        if not g['text'].strip():
            continue
        ty = D[j, 1]
        lo, hi = np.searchsorted(Py, ty - args.window), np.searchsorted(Py, ty + args.window)
        if hi <= lo:
            continue
        cand = order[lo:hi]
        dd = np.abs(Pxy[cand] - D[j])
        ok = np.where(dd[:, 0] < args.window)[0]
        if len(ok) == 0:
            continue
        k = ok[np.argmin(dd[ok].sum(1))]
        matched.append((j, int(cand[k]), float(Pxy[cand[k], 0] - D[j, 0]), float(Pxy[cand[k], 1] - D[j, 1])))
    n_ink = sum(1 for g in G if g['text'].strip())
    print(f'matched {len(matched)} of {n_ink} viewer glyphs within {args.window} pt')
    if not matched:
        sys.exit('nothing matched: are the two the same document at the same width?')
    M = np.array([(m[2], m[3]) for m in matched])
    ys = np.array([G[m[0]]['y'] for m in matched])

    # vertical: the drift, as the median dy per 250 pt of height
    edges = np.arange(0, ys.max() + 250, 250)
    drift = []
    for e0, e1 in zip(edges, edges[1:]):
        s = (ys >= e0) & (ys < e1)
        drift.append(round(float(np.median(M[s, 1])), 3) if s.sum() > 10 else None)
    print(f'vertical (strip − viewer, pt): mean {M[:, 1].mean():+.3f}, sd {M[:, 1].std():.3f}, '
          f'largest {np.abs(M[:, 1]).max():.3f}; drift per 250 pt: {[d for d in drift if d is not None][:8]} … {[d for d in drift if d is not None][-3:]}')

    # horizontal, per line (glyphs sharing a baseline): offset at the start, slope along the line
    lines = collections.defaultdict(list)
    for (j, i, dx, dy) in matched:
        lines[round(G[j]['y'], 2)].append((D[j, 0], dx, G[j]['text']))
    table = []
    for y, pts in sorted(lines.items()):
        pts.sort()
        x = np.array([p[0] for p in pts]); dx = np.array([p[1] for p in pts])
        slope = float(np.polyfit(x, dx, 1)[0] * 100) if len(pts) >= 6 and x.max() - x.min() > 50 else 0.0
        table.append({'y': y, 'glyphs': len(pts), 'start_dx': round(float(dx[0]), 3), 'end_dx': round(float(dx[-1]), 3),
                      'max_abs_dx': round(float(np.abs(dx).max()), 3), 'slope_pt_per_100pt': round(slope, 3),
                      'text': ''.join(p[2] for p in pts)[:60]})
    off = [t for t in table if t['max_abs_dx'] > args.line_tol]
    print(f'horizontal: {len(table)} lines, {len(off)} with a glyph off by more than {args.line_tol} pt')
    for t in sorted(off, key=lambda t: -t['max_abs_dx'])[:25]:
        print(f'  y {t["y"]:9.2f}  start {t["start_dx"]:+.2f}  end {t["end_dx"]:+.2f}  slope {t["slope_pt_per_100pt"]:+.3f}/100pt  {t["text"]!r}')

    # rules: each viewer rect against the strip rule whose corners are
    # nearest. `off` is the farthest any corner of either is from the
    # other's nearest corner (pt): position, size and angle in one number.
    # dx/dy are the move between their centres (strip − viewer).
    def corner_distance(A, B):
        dd = np.hypot(*(A[:, None, :] - B[None, :, :]).transpose(2, 0, 1))
        return max(dd.min(1).max(), dd.min(0).max())
    centres = np.array([q.mean(0) for q in rules]) if rules else np.zeros((0, 2))
    rule_rows, drawn = [], set()
    for r in dom['rects']:
        if r['w'] <= 0 or r['h'] <= 0:
            continue
        V = np.array(r['pts']) + [shift, 0]
        c = V.mean(0)
        near = np.where(np.abs(centres - c).max(1) < 2 + np.ptp(V, 0).max() / 2)[0] if len(rules) else []
        best = min(((corner_distance(V, rules[k]), k) for k in near), default=None)
        if best and best[0] < 2:
            off, k = best
            drawn.add(k)
            rule_rows.append({'y': round(r['y'], 2), 'x': round(r['x'] + shift, 2), 'off': round(float(off), 3),
                              'dx': round(float(centres[k, 0] - c[0]), 3), 'dy': round(float(centres[k, 1] - c[1]), 3)})
        else:
            rule_rows.append({'y': round(r['y'], 2), 'unmatched': True, 'w': round(r['w'], 2), 'h': round(r['h'], 2)})
    # and the rules TeX drew that the browser did not
    missing = [{'y': round(float(rules[k][:, 1].min()), 2), 'x': round(float(rules[k][:, 0].min()), 2)}
               for k in range(len(rules)) if k not in drawn]
    rm = [r for r in rule_rows if 'dx' in r]
    if rm:
        a = np.array([(r['dx'], r['dy']) for r in rm])
        print(f'rules: {len(rm)} matched of {len(rule_rows)} drawn, {len(missing)} of TeX\'s not drawn; '
              f'strip − viewer x sd {a[:, 0].std():.3f}, y sd {a[:, 1].std():.3f}, '
              f'largest corner distance {max(r["off"] for r in rm):.3f} pt')
    for r in sorted((r for r in rm if r['off'] > args.line_tol), key=lambda r: -r['off'])[:10]:
        print(f'  rule at y {r["y"]:9.2f}: corners {r["off"]:.2f} pt off, centre dx {r["dx"]:+.2f} dy {r["dy"]:+.2f} pt')

    (out / 'vector.json').write_text(json.dumps({
        'hsize_pt': hsize, 'margin_pt': margin, 'window_pt': args.window,
        'glyphs': {'strip': len(P), 'viewer': n_ink, 'matched': len(matched)},
        'vertical': {'mean': round(float(M[:, 1].mean()), 4), 'sd': round(float(M[:, 1].std()), 4),
                     'max_abs': round(float(np.abs(M[:, 1]).max()), 4), 'drift_per_250pt': drift},
        'lines': table, 'lines_off': off, 'rules': rule_rows, 'rules_missing': missing,
        'matched': [{'y': round(G[j]['y'], 2), 'x': round(D[j, 0], 2), 'text': G[j]['text'], 'font': G[j]['font'],
                     'dx': round(dx, 3), 'dy': round(dy, 3)} for (j, i, dx, dy) in matched],
    }, indent=1))
    print(f'→ {out / "vector.json"}')


if __name__ == '__main__':
    main()
